### PR TITLE
python312Packages.pyside6: allow optional dependencies for darwin

### PR DESCRIPTION
**Pull request by gador**
## Description of changes

This PR refactors `pyside6` by:

- adding `LLVM_INSTALL_DIR` environment variable to let `pyside6` find the `libclang` library
- add required darwin frameworks
- use a `symlinkJoin` construct of all qt6 dependencies so `pyside6` finds the dependencies on darwin. This is necessary, because `pyside6` on macos expects all dependencies to be under the same directory
- remove the optional module `Designer` from the darwin python bindings, since it causes linker failures.
- adding a `pythonImportsCheck`

The benefit of this change is that `pyside6` now includes almost all the python bindings for Qt (except Designer and qtwebengine) on darwin.

This also adds the following modules to the `x86_64-linux` binding:

- qtpositioning
- qtlocation
- qtshadertools
- qtserialport
- qtserialbus
- qtgraphs

Ping @gebner 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gador** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 329238` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>comic-mandown: <a href="https://termbin.com/k9xq">plain log</a></li>
  </ul>
</details>
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>comic-mandown.dist: <a href="https://termbin.com/nbov">plain log</a></li>
  </ul>
</details>
<details>
<summary>
29 packages built:
</summary>
<ul>
<li>cutter</li>
<li>cutterPlugins.rz-ghidra</li>
<li>flashgbx</li>
<li>flashgbx.dist</li>
<li>kdePackages.falkon</li>
<li>kdePackages.falkon.debug</li>
<li>kdePackages.falkon.dev</li>
<li>kdePackages.falkon.devtools</li>
<li>nitrokey-app2</li>
<li>nitrokey-app2.dist</li>
<li>normcap</li>
<li>normcap.dist</li>
<li>onedrivegui</li>
<li>onedrivegui.dist</li>
<li>onionshare</li>
<li>onionshare-gui</li>
<li>onionshare-gui.dist</li>
<li>onionshare.dist</li>
<li>openusd</li>
<li>python311Packages.pyside6</li>
<li>python312Packages.pyside6</li>
<li>retool</li>
<li>retool.dist</li>
<li>smb3-foundry</li>
<li>streamdeck-ui</li>
<li>streamdeck-ui.dist</li>
<li>syncplay</li>
<li>typstwriter</li>
<li>typstwriter.dist</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
cutter:
<details>

got build log for '/nix/store/7iawsv10d4mb4inzw8g4dk14nq0g9bhp-cutter-2.3.4' from 'daemon'
When evaluating attribute ‘cutter’:
warning: unclear-gpl
`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/development/tools/analysis/rizin/cutter.nix:88:5:

```
   |
88 |     license = licenses.gpl3;
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
cutterPlugins.rz-ghidra:
<details>

got build log for '/nix/store/4g65cg485d8wsgr1gxqnnsr55cjbblx9-rz-ghidra-0.7.0' from 'daemon'
When evaluating attribute ‘cutterPlugins.rz-ghidra’:
warning: unclear-gpl
`lgpl3` is a deprecated license, please check if project uses `lgpl3Plus` or `lgpl3Only` and change `meta.license` accordingly.

Near pkgs/development/tools/analysis/rizin/rz-ghidra.nix:56:5:

```
   |
56 |     license = licenses.lgpl3;
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
</details>

**Comment by gador** [comment]
Result of `nixpkgs-review pr 329238` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>cutter</li>
    <li>cutterPlugins.rz-ghidra</li>
    <li>normcap</li>
    <li>normcap.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>comic-mandown</li>
    <li>comic-mandown.dist</li>
  </ul>
</details>
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>flashgbx</li>
    <li>flashgbx.dist</li>
    <li>nitrokey-app2</li>
    <li>nitrokey-app2.dist</li>
    <li>onionshare</li>
    <li>onionshare.dist</li>
    <li>openusd</li>
    <li>python311Packages.pyside6</li>
    <li>python312Packages.pyside6</li>
    <li>retool</li>
    <li>retool.dist</li>
    <li>smb3-foundry</li>
    <li>syncplay</li>
    <li>typstwriter</li>
    <li>typstwriter.dist</li>
  </ul>
</details>

**Comment by gador** [comment]
Result of `nixpkgs-review pr 329238` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>cutter</li>
    <li>cutterPlugins.rz-ghidra</li>
    <li>normcap</li>
    <li>normcap.dist</li>
    <li>onionshare-gui</li>
    <li>onionshare-gui.dist</li>
    <li>streamdeck-ui</li>
    <li>streamdeck-ui.dist</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>comic-mandown</li>
    <li>comic-mandown.dist</li>
    <li>openusd</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>flashgbx</li>
    <li>flashgbx.dist</li>
    <li>nitrokey-app2</li>
    <li>nitrokey-app2.dist</li>
    <li>onionshare</li>
    <li>onionshare.dist</li>
    <li>python311Packages.pyside6</li>
    <li>python312Packages.pyside6</li>
    <li>retool</li>
    <li>retool.dist</li>
    <li>smb3-foundry</li>
    <li>syncplay</li>
    <li>typstwriter</li>
    <li>typstwriter.dist</li>
  </ul>
</details>

**Comment by gador** [comment]
`openusd` fails on `x86_64-darwin` on `master` also, due to too old macOS SDK

File: pkgs/development/python-modules/pyside6/default.nix [hunk 1]
```diff
@@ -15,52 +58,55 @@ stdenv.mkDerivation (finalAttrs: {
...
+    '';
+
+  # "Couldn't find libclang.dylib You will likely need to add it manually to PATH to ensure the build succeeds."
+  env = lib.optionalAttrs stdenv.isDarwin { LLVM_INSTALL_DIR = "${llvmPackages.libclang.lib}/lib"; };
 
   nativeBuildInputs = [
     cmake
     ninja
     python
+    python.pkgs.pythonImportsCheckHook
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyside6/default.nix:86]:
```suggestion
    pythonImportsCheckHook
```

**Comment by reckenrode** [comment]
> remove the optional module Designer from the darwin python bindings, since it causes linker failures.

What are these linker failures? The linker was just upgraded in master, so it’s possible the issue may have been resolved.

**Comment by gador** [comment]
> > remove the optional module Designer from the darwin python bindings, since it causes linker failures.
> 
> What are these linker failures? The linker was just upgraded in master, so it’s possible the issue may have been resolved.

```pyside6> qt.shiboken: (designer) ~FileOut file /nix/store/yn6n8d8hh3id5wc5zlhy4vz2hzjn321b-pyside6-6.7.2/.build/pyside-setup-everywhere-src-6.7.2/sources/pyside6/build/PySide6/QtDesigner/PySide6/QtDesigner/qpydesignercustomwidgetcollection_wrapper.cpp not written.
pyside6> shiboken error: Could not find class "QDesignerCustomWidgetInterface" in the code model. Maybe it is forward declared?
```
Is the error on a rebased on current `master` version. It is the same I encountered. `QDesignerCustomWidgetInterface` found in `qttools`:
```
nix-locate QDesignerCustomWidgetInterface
qt6.qttools.out                                      26 r /nix/store/1gyz3xyfchyi15x98r1c6ypg4lpjnvai-qttools-6.5.3/include/QtUiPlugin/QDesignerCustomWidgetInterface
```

So it should be found, but is not.

I'm also not too sure how important `QtDesigner` is for python bindings. Also, as `qttools` was not included in `darwin` before, `QtDesigner` wasn't included anyway until I added `qttools` for `darwin`.

**Comment by reckenrode** [comment]
Thanks. That’s looks like a shiboken issue, which would presumably be unrelated to whatever ld is doing.

**Pull request merged**